### PR TITLE
Add logs around SSR cache accesses

### PR DESCRIPTION
**Pull request by noahtallen**
### Proposed Changes
My ultimate goal is to build a Grafana dashboard around the caches related to the SSR pipeline. This will help us answer if cache is effectively utilized for most requests.

I'm building this on top of my statsd changes in #68068.

Each log follows the same format and uses a statsd counter. We log the total number of hits and the total number of misses, which summed together equal the total number of accesses.

```js
logServerEvent( sectionName, {
	name: `ssr.name_of_cache.${ dataExists ? 'hit' : 'miss' }`,
	type: 'counting',
} );
```

I added this in the following places:
- Themes fetch cache access. (Which happens during SSR of the themes page)
- Server redux cache. (Happens on every request)
- React SSR markup cache. (Happens every time we render SSR markup)

I also looked into adding this to /plugins, around the server-side react query stuff, but realized it would be more tricky than this, since there isn't a single place to check if the cache was accessed and missed as part of a fetch. (Open to suggestions on how to handle this)

#### Potential concerns:
This should add several extra fetch requests to the server with `superagent`, since we aren't throttling anything client side. I would prefer to not throttle anything, since there should be a level of throttling on the other end of the system anyway (in statsd or graphite or boom.gif). We just need to make sure this doesn't impact the server load itself. I don't _think_ it should, since the server never needs to wait for the `get` request to finish, but I wonder if there are any ways to optimize this.

### Testing instructions:
TBD. Not quite sure how to test this, other than smoke testing SSR.

### CODE REVIEW

File: client/lib/analytics/statsd-utils.ts [hunk 1]
```diff
@@ -49,13 +49,13 @@ export function createStatsdURL( calypsoSection: string, events: BeaconData[] |
...
  * and can disable statsd events with the server_side_boom_analytics_enabled
- * config flag.
+ * config flag. Does nothing if called on the client.
  *
  * @param calypsoSection The Calypso section the event occurred under.
  * @param events List of events to send to the server.
  */
 export function logServerEvent( calypsoSection: string, events: BeaconData[] | BeaconData ) {
-	if ( config( 'server_side_boom_analytics_enabled' ) ) {
+	if ( typeof window === 'undefined' && config( 'server_side_boom_analytics_enabled' ) ) {
```
Comment (noahtallen) [inline client/lib/analytics/statsd-utils.ts:75]: Some of the theme fetch functions can also be called on the client. For now, we just want to see server stats here. (Noting that we currently have no statsd events on the client anyways.)

**Review by tyxla — APPROVED** [review]
Looks good, thanks 👍

File: client/server/pages/index.js [hunk 2]
```diff
@@ -123,6 +124,13 @@ function getDefaultContext( request, response, entrypoint = 'entry-main' ) {
 	const cachedServerState = request.context.isLoggedIn ? {} : stateCache.get( cacheKey ) || {};
 	const getCachedState = ( reducer, storageKey ) => {
 		const storedState = cachedServerState[ storageKey ];
+
+		logServerEvent( request.context.sectionName, {
```
Comment (claudiucelfilip) [inline client/server/pages/index.js]: The `request.context.sectionName` is `undefined` here, probably because this is [called](https://github.com/Automattic/wp-calypso/blob/e538d107e7fc0741e64604797f2fec642bf2c5e3/client/server/pages/index.js#L820) before [setUpSectionContext](https://github.com/Automattic/wp-calypso/blob/e538d107e7fc0741e64604797f2fec642bf2c5e3/client/server/pages/index.js#L524).
This triggers an error [here](https://github.com/Automattic/wp-calypso/blob/e538d107e7fc0741e64604797f2fec642bf2c5e3/client/lib/analytics/statsd-utils.ts#L36) and I expect the e2e are failing because of this.

Comment (noahtallen) [reply client/server/pages/index.js]: Thanks for looking into this!

**Comment by noahtallen** [comment]
Add two improvements around the section name:

1. Since `setupDefaultContext` doesn't have access to the section yet, we pass it the section name. (Side-note: this fixes _another_ case in `getDefaultContext` where we were using context.sectionName, even though it was undefined.)
2. If the calypso section is undefined when we get to "log server event", we just set it to unknown

File: client/lib/analytics/statsd-utils.ts [hunk 3]
```diff
@@ -65,13 +65,14 @@ export function createStatsdURL( calypsoSection: string, events: BeaconData[] |
...
  * and can disable statsd events with the server_side_boom_analytics_enabled
- * config flag.
+ * config flag. Does nothing if called on the client.
  *
  * @param calypsoSection The Calypso section the event occurred under.
  * @param events List of events to send to the server.
  */
 export function logServerEvent( calypsoSection: string, events: BeaconData[] | BeaconData ) {
-	if ( config( 'server_side_boom_analytics_enabled' ) ) {
+	calypsoSection ??= 'unknown';
```
Comment (noahtallen) [inline client/lib/analytics/statsd-utils.ts:74]: This should likely never get triggered, but it's preferable to a crash since we don't have strict types everywhere on the server yet.

File: client/server/pages/index.js [hunk 4]
```diff
@@ -140,7 +148,7 @@ function getDefaultContext( request, response, entrypoint = 'entry-main' ) {
 
 	const oauthClientId = request.query.oauth2_client_id || request.query.client_id;
 	const isWCComConnect =
-		( 'login' === request.context.sectionName || 'signup' === request.context.sectionName ) &&
+		( 'login' === sectionName || 'signup' === sectionName ) &&
```
Comment (noahtallen) [inline client/server/pages/index.js:151]: Drive-by-fix: context.sectionName is undefined everywhere in this function, so there's no way this was working.

File: client/server/pages/index.js [hunk 5]
```diff
@@ -95,7 +96,7 @@ function setupLoggedInContext( req, res, next ) {
 	next();
 }
 
-function getDefaultContext( request, response, entrypoint = 'entry-main' ) {
+function getDefaultContext( request, response, entrypoint = 'entry-main', sectionName ) {
```
Comment (noahtallen) [inline client/server/pages/index.js:99]: entrypoint appears to always be passed to this function.
